pkg/util/steps/runner.go

```diff
@@ -0,0 +1,177 @@
+package steps
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/Azure/ARO-RP/pkg/util/log"
+)
+
+// StepResultHandler is the interface for the handler of Runner.
+//
+// OnStepError is called when a step or condition fails and returns the error to
+// fail with. This error may be different than the error passed in originally.
+//
+// ShouldRetryStep is called when an Action fails to check whether it should be
+// retried. A return value of (true, nil) indicates it should be reran, all
+// other return values will cause the Action to fail. If the check itself causes
+// an error, it may return that instead, but it should also log the original
+// error. ShouldRetryStep will be ran in the same timeout deadline as the step
+// that failed.
+type StepResultHandler interface {
+	OnStepError(ctx context.Context, err error, step Step) error
+	ShouldRetryStep(ctx context.Context, err error, step ActionFunction) (bool, error)
+}
+
+// Runner is a runner of task steps and checker of conditions.
+type Runner struct {
+	name         string
+	log          *logrus.Entry
+	steps        []Step
+	pollInterval time.Duration
```

Comment:
Suggest pulling the pollInterval const into this package and not making it configurable.  We can make it configurable later if it turns out that we need to (I think that's unlikely).